avocado/plugins/beaker_result.py

```diff
@@ -0,0 +1,137 @@
+"""
+beaker result plugin
+
+Sends results and logs to beaker using the harness API.
+https://beaker-project.org/docs/alternative-harnesses/index.html
+
+Finds the beaker API entry point using the BEAKER_LAB_CONTROLLER_URL
+environment variable.  Does nothing in case the variable is not set.
+
+"""
+import glob
+import os
+import pprint
+import urllib.error
+import urllib.parse
+import urllib.request
+
+from avocado.core.output import LOG_UI
+from avocado.core.plugin_interfaces import ResultEvents
+
+
+class BeakerResult(ResultEvents):
+
+    """
+    send test results to beaker test harness api
+    """
+
+    name = 'beaker'
+    description = 'report results to beaker'
+
+    beaker_url = None
+    job_id = None
+
+    def __init__(self, config):
+        self.config = config  # make lint happy
+        baseurl = os.environ.get('BEAKER_LAB_CONTROLLER_URL')
+        recipeid = os.environ.get('RSTRNT_RECIPEID')
+        taskid = os.environ.get('RSTRNT_TASKID')
+        if baseurl is None or recipeid is None or taskid is None:
+            return
+        baseurl = baseurl.rstrip('/')
+        self.beaker_url = baseurl + '/recipes/' + recipeid + '/tasks/' + taskid
+        LOG_UI.info("beaker: using API at %s (R:%s T:%s)", baseurl, recipeid, taskid)
+
+    def send_request(self, req):
+        LOG_UI.info('beaker: %s %s ...', req.method, req.full_url)
+        try:
+            res = urllib.request.urlopen(req)  # nosec
+            return res
+        except urllib.error.URLError as err:
+            LOG_UI.info('beaker: Failed: %s', err)
+            return None
+        except Exception as err:
+            # should not happen
+            LOG_UI.info('beaker: Oops: %s', err)
+            return None
+
+    def post_result(self, state):
+        reqdict = {
+            'path': "%s/%s" % (self.job_id, state.get('name')),
+        }
+
+        result = state.get('status').lower().capitalize()
+        if result == 'Cancel':
+            result = 'Skip'
+        if result not in ['Pass', 'Fail', 'Skip']:
+            result = 'Fail'
+        reqdict['result'] = result
+
+        reason = state.get('fail_reason')
+        if reason is not None:
+            reqdict['message'] = reason
+
+        secs = state.get('time_elapsed')
+        if secs is not None:
+            reqdict['score'] = str(int(secs))
```

Comment:
I see the `score` is a free form field... I wonder if, choosing to use it for the time, it makes sense to give it more context such as a string stating: "Duration: xx seconds".

But I do not know what other harnesses do, or what were your original goals here, so this may be a bad suggestion.

Reply:
It's an integer not a string (even though I have to convert it to a string so I can feed it into urllib.parse.urlencode).
Documentation explicitly states this has not some specific meaning and free to use to fill in whatever makes sense for the test.  Performance tests might fill in some kind of (test-specific) performance score.

Given avocado hasn't something similiar filling in the duration looked most useful to me.  We could also just leave it blank, or add such a score to avocado.

Reply:
Let's keep it as is then.